pkg/kv/kvtest/store.go

```diff
@@ -0,0 +1,269 @@
+)
+
+type makeStore func(t *testing.T, ctx context.Context) kv.Store
+
+func TestDriver(t *testing.T, name string, dsn string) {
+	ms := makeStoreByName(name, dsn)
+	t.Run("Driver_Open", func(t *testing.T) { testDriverOpen(t, ms) })
+	t.Run("Store_SetGet", func(t *testing.T) { testStoreSetGet(t, ms) })
+	t.Run("Store_SetIf", func(t *testing.T) { testStoreSetIf(t, ms) })
+	t.Run("Store_Delete", func(t *testing.T) { testStoreDelete(t, ms) })
+	t.Run("Store_Scan", func(t *testing.T) { testStoreScan(t, ms) })
+}
+
+func testDriverOpen(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store1 := ms(t, ctx)
+	store1.Close()
+
+	store2 := ms(t, ctx)
+	store2.Close()
+}
+
+func testStoreSetGet(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	testKey := []byte("key")
+	testValue1 := []byte("value1")
+	testValue2 := []byte("value2")
+
+	// set test key with value1
+	err := store.Set(ctx, testKey, testValue1)
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue1, err)
+	}
+
+	// get test key with value1
+	val, err := store.Get(ctx, testKey)
+	switch {
+	case err != nil:
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	case val == nil:
+		t.Fatalf("got value with nil")
+	case !bytes.Equal(testValue1, val):
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val, testValue1)
+	}
+
+	// override key with value2
+	err = store.Set(ctx, testKey, testValue2)
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue2, err)
+	}
+
+	// get test key with value2
+	val2, err := store.Get(ctx, testKey)
+	if err != nil {
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	}
+	if !bytes.Equal(testValue2, val2) {
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val2, testValue2)
+	}
+
+	// get a missing key
+	const keyNotExists = "key-not-exists"
+	val3, err := store.Get(ctx, []byte(keyNotExists))
+	if !errors.Is(err, kv.ErrNotFound) {
+		t.Fatalf("get key='%s' err=%s, expected not found", keyNotExists, err)
+	}
+	if val3 != nil {
+		t.Fatalf("get key='%s' value='%s', expected nil", keyNotExists, val3)
+	}
+}
+
+func testStoreDelete(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("exists", func(t *testing.T) {
+		keyToDel := []byte("key-to-delete")
+		err := store.Set(ctx, keyToDel, []byte("value to delete"))
+		if err != nil {
+			t.Fatalf("failed to set key='%s': %s", keyToDel, err)
+		}
+		err = store.Delete(ctx, keyToDel)
+		if err != nil {
+			t.Fatalf("failed to delete key='%s': %s", keyToDel, err)
+		}
+	})
+
+	t.Run("non_exists", func(t *testing.T) {
+		keyToDel := []byte("missing-key-to-delete")
+		err := store.Delete(ctx, keyToDel)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("delete missing key '%s', err=%v expected=%s", keyToDel, err, kv.ErrNotFound)
+		}
```

Comment:
Do we need to require this behaviour of our KV stores?  It might not be natively supported by some KV stores.  E.g. DynamoDB does not easily support this AFAIK: the [manual says](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteItem.html):

> Unless you specify conditions, the DeleteItem is an idempotent operation; running it multiple times on the same item or attribute does not result in an error response.

Reply:
True - we can simulate it because I'm sure not all our condition will be 1:1 with the underlying features.
Strongly agree in this case - I'll align the implementation.